torchdynamo/mutation_guard.py

```diff
@@ -88,15 +78,24 @@ def is_dynamic_nn_module(obj):
     )
 
 
-def install_generation_tagging_new():
+def install_generation_tagging_init():
     """
-    Monkey patch torch.nn.Module.__new__ so we can detect nn.Module
-    instances created dynamically inside forward methods.
+    Monkey patch torch.nn.Module.__init__ and torch.nn.Module.__setstate__
+    so we can detect nn.Module instances created dynamically inside forward methods.
     """
-    assert (
-        Module.__dict__.get("__new__") is generation_tagging_new
-        or Module.__new__ is object.__new__
-    )
-    Module.__new__ = generation_tagging_new
-    GenerationTracker.db.clear()
+    init = Module.__init__
+
+    def patched_init(self, *args, **kwargs):
```

Comment:
Jason and I agreed that we think there's a fast-path for calls to an unmodified new() in CPython. It isn't monkey-patching itself that's causing the slowdown (that's only done once when we enter the torchdynamo context).

For two vs one method, we need to patch two methods now because when you unpickle, it calls new() to allocate, but doesn't call init(), it calls setstate(). This impl makes more sense anyway, because rather than having some map floating somewhere with id(obj) as the key mapping to the tag, we just put an attribute for the tag directly on the object.